pageserver/src/tenant/mgr.rs

```diff
@@ -1085,6 +1296,7 @@ pub(crate) async fn list_tenants() -> Result<Vec<(TenantId, TenantState)>, Tenan
         .filter_map(|(id, tenant)| match tenant {
             TenantSlot::Attached(tenant) => Some((*id, tenant.current_state())),
             TenantSlot::Secondary => None,
+            TenantSlot::InProgress(_) => None,
```

Comment:
This seems dangerous. A tenant that is being worked on won't show up in the listing. Can't have that IMO. We should extend the api spec to allow returning a different body than full `TenantState` for tenants that are `InProgress`. Maybe a string that's provided by the acquirer of the slot, indicating why the tenant is `InProgres`. That'd also be a good debugging breadcrumb.

Reply:
I understand that this is surprising on first glance, but I don't think it is dangerous.

Remember that when we're in this state, we'are always transitioning to or from a state where the tenant would not be visible in the list API:
- while a request to attach a tenant is in progress, we make no promises about that tenant showing up in the external API until we have acked that attach request.
- when we get a request to detach a tenant (or transition to secondary), we do not guarantee the visibility of the tenant throughout that request: there will be some moment in time when that request is still happening (we haven't sent a response), but the tenant is already not present in the listing.

So the API semantic is unchanged:
- while waiting for the response to an attach/detach request, the tenant _might_ be present in the listing
- once the request has been acked, the tenant will be definitively present or absent.

Reply:
Ok, I agree that it's not a semantics change for the existing APIs.

But, it'll be quite odd for `upsert`: when doing upsert in the not-fast-pasth, the tenant briefly disappears.

That being said, I think we should review who is actually using the list endpoint.
I don't think console does.
We should de-publish it from the openapi spec, then I'm not worried anymore.

We should also reduce the published API surface for the per-tenant status endpoint.

Can you do this as a preliminary to this PR? The way to figure out if there's a dependency on the API surface is to make the change in cloud.git first, and if it still compiles, commit the change there, then mirror it in neon.git.

Reply:
https://neondb.slack.com/archives/C033RQ5SPDH/p1699021103513689?thread_ts=1699013959.869989&cid=C033RQ5SPDH